heron/scheduler-core/src/java/com/twitter/heron/scheduler/SubmitterMain.java

```diff
@@ -389,80 +403,60 @@ public boolean submitTopology() {
       // TODO(mfu): timeout should read from config
       SchedulerStateManagerAdaptor adaptor = new SchedulerStateManagerAdaptor(statemgr, 5000);
 
-      boolean isValid = validateSubmit(adaptor, topology.getName());
+      validateSubmit(adaptor, topology.getName());
 
       // 2. Try to submit topology if valid
-      if (isValid) {
-        // invoke method to submit the topology
-        LOG.log(Level.FINE, "Topology {0} to be submitted", topology.getName());
-
-        // Firstly, try to upload necessary packages
-        packageURI = uploadPackage(uploader);
-        if (packageURI == null) {
-          LOG.severe("Failed to upload package.");
-          return false;
-        } else {
-          // Secondly, try to submit a topology
-          // build the runtime config
-          Config runtime = Config.newBuilder()
-              .putAll(LauncherUtils.getInstance().getPrimaryRuntime(topology, adaptor))
-              .put(Keys.topologyPackageUri(), packageURI)
-              .put(Keys.launcherClassInstance(), launcher)
-              .build();
-
-          isSuccessful = callLauncherRunner(runtime);
-        }
-      }
+      // invoke method to submit the topology
+      LOG.log(Level.FINE, "Topology {0} to be submitted", topology.getName());
+
+      // Firstly, try to upload necessary packages
+      URI packageURI = uploadPackage(uploader);
+
+      // Secondly, try to submit a topology
+      // build the runtime config
+      Config runtime = Config.newBuilder()
+          .putAll(LauncherUtils.getInstance().getPrimaryRuntime(topology, adaptor))
+          .put(Keys.topologyPackageUri(), packageURI)
+          .put(Keys.launcherClassInstance(), launcher)
+          .build();
+
+      callLauncherRunner(runtime);
+    } catch (LauncherException | PackingException e) {
+      uploader.undo();
```

Comment:
should explain in comments why you only call undo for certain exceptions. Or better yet, could we always call undo and have that method be written in a way where it's a noop if there is nothing for it to do?